### ISSUE TITLE
createWorkspace fails on self-hosted deploy with private CA

### ISSUE THREAD

**Issue by dpb587** [gitpod-io/gitpod#1434]
Hi – any pointers would be appreciated. Thanks!

### Describe the bug

When attempting to create a new workspace in a self-hosted environment (with private CA), it fails with the following error in the UI.

 > Error: Request createWorkspace failed with message: 13 INTERNAL: cannot resolve workspace image: Error response from daemon: Get https://reg.gitpod.acme.net/v2/: x509: certificate signed by unknown authority

And the following appears in the `server` logs.

```json
{
  "@type": "type.googleapis.com/google.devtools.clouderrorreporting.v1beta1.ReportedErrorEvent",
  "serviceContext": {
    "service": "server",
    "version": "v0.3.0"
  },
  "stack_trace": "Error: 13 INTERNAL: cannot resolve workspace image: Error response from daemon: Get https://reg.gitpod.acme.net/v2/: x509: certificate signed by unknown authority\n    at Object.exports.createStatusError (/app/node_modules/grpc/src/common.js:91:15)\n    at Object.onReceiveStatus (/app/node_modules/grpc/src/client_interceptors.js:1204:28)\n    at InterceptingListener._callNext (/app/node_modules/grpc/src/client_interceptors.js:568:42)\n    at InterceptingListener.onReceiveStatus (/app/node_modules/grpc/src/client_interceptors.js:618:8)\n    at callback (/app/node_modules/grpc/src/client_interceptors.js:845:24)",
  "component": "server",
  "severity": "ERROR",
  "time": "2020-04-23T02:55:50.923Z",
  "environment": "production",
  "region": "local",
  "message": "Request createWorkspace failed with internal server error",
  "error": "Error: 13 INTERNAL: cannot resolve workspace image: Error response from daemon: Get https://reg.gitpod.acme.net/v2/: x509: certificate signed by unknown authority\n    at Object.exports.createStatusError (/app/node_modules/grpc/src/common.js:91:15)\n    at Object.onReceiveStatus (/app/node_modules/grpc/src/client_interceptors.js:1204:28)\n    at InterceptingListener._callNext (/app/node_modules/grpc/src/client_interceptors.js:568:42)\n    at InterceptingListener.onReceiveStatus (/app/node_modules/grpc/src/client_interceptors.js:618:8)\n    at callback (/app/node_modules/grpc/src/client_interceptors.js:845:24)",
  "payload": {
    "method": "createWorkspace",
    "args": [
      {
        "contextUrl": "https://github.com/dpb587/gget",
        "mode": "default"
      },
      {
        "_isCancelled": false
      }
    ]
  }
}
```

Additionally, I experimented with the configuring the following while trying to resolve the error, unsuccessfully:

 * configuring `gitpod.components.server.*` with custom ca-certificates
 * configuring `gitpod.docker-registry.tlsSecretName` with a managed cert

### Steps to reproduce

Deploy `self-hosted` with https + self-signed certs (via `certstrap`). Generally...

```
$ echo values/https.yaml >> configuration.txt
$ mkdir secrets/https-certificates pki
$ cd pki
$ certstrap init --common-name=gitpod-ca --passphrase ''
$ certstrap request-cert --common-name=gitpod-proxy --domain=gitpod.acme.net,*.gitpod.acme.net,*.ws.gitpod.acme.net --passphrase ''
$ certstrap sign --CA gitpod-ca --csr out/gitpod-proxy.csr gitpod-proxy
$ cp out/gitpod-proxy.crt ../secrets/https-certificates/cert.pem
$ cp out/gitpod-proxy.key ../secrets/https-certificates/privkey.pem
$ cat out/gitpod-{proxy,ca}.crt > ../secrets/https-certificates/fullchain.pem
$ cat out/gitpod-{proxy,ca}.crt > ../secrets/https-certificates/chain.pem
$ cd ..
```

### Expected behavior

Workspace to be created.

### Additional information

I manually tried verifying the connection via `exec` on the server pod.

```
$ curl https://reg.gitpod.acme.net/v2/
curl: (60) SSL certificate problem: self signed certificate in certificate chain
```

I attempted to merge in the following configuration which bundled my private CA, but it didn't seem to help.

```yaml
gitpod:
  components:
    server:
      serverContainer:
        volumeMounts:
        - name: ca-certificates
          mountPath: /etc/ssl/certs
          readOnly: true
      volumes:
      - name: ca-certificates
        configMap:
          name: ca-certificates
```

Although it did noticeably fix it for default requests like `curl`, at least.

```
$ curl https://reg.gitpod.acme.net/v2/
<html>
<head><title>401 Authorization Required</title></head>
<body bgcolor="white">
<center><h1>401 Authorization Required</h1></center>
<hr><center>nginx/1.14.1</center>
</body>
</html>
```

I also tried overriding the docker certs with managed certificates (and tried switching its `service.port` to `80`), but it didn't seem to change much.

```yaml
gitpod:
  docker-registry:
    tlsSecretName: alternative-registry-certs
```

**Comment by meysholdt** [comment]
hi @dpb587, that's quite a detailed report. Thank you!

The `reg.gitpod.acme.net` points to Gitpod's service named `proxy`, which does https-termination and http-basic-auth. The `proxy` then forwards the requests to an internal docker registry. You can also [configure an external docker registry](https://www.gitpod.io/docs/self-hosted/latest/install/docker-registry/). Gitpod uses this image registry to store docker images it has built itself.

`reg.gitpod.acme.net` is being accessed by:
* the `kubelet`/`containerd`/`dockerd` from you Kubernetes installation
* Gitpod's `server` component
* Gitpod's `image-builder` component. 

Things you can do for troubleshooting:
* check if your cert has been configured in the `proxy` pod.
* check if you can `docker push` / `docker pull` to `reg.gitpod.acme.net`. E.g. run
```
docker pull alpine
docker tag alpine reg.gitpod.acme.net/alpine:latest
docker push reg.gitpod.acme.net/alpine:latest
```
* if the above works and you get no cert error with `docker push` or `curl`, but the `server` still shows the cert error, then it may make sense to check if the certificate authority of your cert is known in the `server` pod.

It may also be interesting to try using [an external docker registry](https://www.gitpod.io/docs/self-hosted/latest/install/docker-registry/). This is a setting we recommend for production usages anyway and it may solve your certificate troubles.

**Comment by dpb587** [comment]
Thanks, @meysholdt! The explanation was very helpful – I was able to get it working with the chart-provided Docker registry.

To document my steps a bit more... I replaced the `builtin` and used a more-properly configured, custom `components.imageBuilder.registryCerts` value where `name` was my hostname of `reg.gitpod.acme.net`.

I configured both `components.server.{serverContainer,volumes}` and `container.configmap.{volumeMounts,volumes}` with my custom-compiled CA file mentioned earlier. Not actually sure if both were needed (haven't automated this yet).

For `kubelet`/`containerd`, I ended up including a section like: 

```toml
[plugins.cri.registry.configs."reg.gitpod.acme.net".tls]
  ca_file = "/etc/docker/certs.d/reg.gitpod.acme.net/ca.crt"
```

Thanks, again – looking forward to evaluating this more.
